File: examples/p3_oop_balls.py
#!/usr/bin/env python3
'''
@author  Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d
from p2_oop_ball import Ball, ARENA_W, ARENA_H

def tick():
    g2d.clear_canvas()  # BG
    for b in balls:
        b.move()
        g2d.fill_rect(b.position())  # FG

def main():
    global balls
    balls = [Ball((40, 80)), Ball((80, 40)), Ball((120, 120))]
    g2d.init_canvas((ARENA_W, ARENA_H))
    g2d.main_loop(tick)

main()

File: go/p3_oop_balls.go
package main

import . "g2d"

var screen = Point{480, 360}
var size = Point{20, 20}

type Ball struct {
    x, y   int
    dx, dy int
}

func NewBall(pos Point) *Ball {
    return &Ball{pos.X, pos.Y, 5, 5}
}

func (b *Ball) Move() {
    if !(0 <= b.x+b.dx && b.x+b.dx <= screen.X-size.X) {
        b.dx = -b.dx
    }
    if !(0 <= b.y+b.dy && b.y+b.dy <= screen.Y-size.Y) {
        b.dy = -b.dy
    }
    b.x += b.dx
    b.y += b.dy
}

func (b *Ball) Pos() Point {
    return Point{b.x, b.y}
}

// Create two objects, instances of the Ball class
var balls = []*Ball{NewBall(Point{40, 80}), NewBall(Point{80, 40})}

func tick() {
    ClearCanvas() // BG
    for _, b := range balls {
        b.Move()
        DrawImage("ball.png", b.Pos()) // FG
    }
}

func main() {
    InitCanvas(screen)
    MainLoop(tick)
}
